compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNotImplementedOverrideChecker.kt

```diff
@@ -0,0 +1,287 @@
+                } else {
+                    reporter.reportOn(source, FirErrors.MANY_IMPL_MEMBER_NOT_IMPLEMENTED, declaration, representative, context)
+                }
+                alreadyReportedManyNotImplemented = true
+*/
+            } else {
+                assert(fakeOverrides.size == 1)
+                val notImplemented = fakeOverrides.single()
+                if ((notImplemented as? FirMemberDeclaration)?.isAbstract != true) continue
+                if (canHaveAbstractMembers || alreadyReportedAbstractNotImplemented) continue
+
+                val notImplementedFunction = notImplemented as? FirSimpleFunction
+                // TODO: suspend function overridden by a Java class in the middle is not properly regarded as an override.
+                if (notImplementedFunction?.isSuspend == true) continue
+                // TODO: Collection extensions are not properly regarded as an override.
+                val originalPackage =
+                    notImplementedFunction?.originalIfFakeOverride()?.dispatchReceiverType?.classId?.packageFqName?.asString()
+                if (originalPackage in COLLECTIONS) continue
+
+                if (notImplemented.isFromInterface(context)) {
+                    reporter.reportOn(source, FirErrors.ABSTRACT_MEMBER_NOT_IMPLEMENTED, declaration, notImplemented, context)
+                } else {
+                    reporter.reportOn(source, FirErrors.ABSTRACT_CLASS_MEMBER_NOT_IMPLEMENTED, declaration, notImplemented, context)
+                }
+                alreadyReportedAbstractNotImplemented = true
+            }
+        }
+    }
+
+    private fun FirCallableDeclaration<*>.isFromInterface(context: CheckerContext): Boolean =
+        (getContainingClass(context) as? FirRegularClass)?.isInterface == true
+
+    private fun collectCallableMembers(firClass: FirClass<*>, context: CheckerContext): Collection<FirCallableDeclaration<*>> =
+        firClass.declarations.filterIsInstance<FirCallableDeclaration<*>>() +
+                dataClassMembers(firClass, context) +
+                delegatedMembers(firClass, context)
+
+    private val FirSimpleFunction.matchesEqualsSignature: Boolean
+        get() = valueParameters.size == 1 && valueParameters[0].returnTypeRef.coneType.isNullableAny &&
+                returnTypeRef.coneType.isBoolean
+
+    private val FirSimpleFunction.matchesHashCodeSignature: Boolean
+        get() = valueParameters.isEmpty() &&
+                returnTypeRef.coneType.isInt
+
+    private val FirSimpleFunction.matchesToStringSignature: Boolean
+        get() = valueParameters.isEmpty() &&
+                returnTypeRef.coneType.isString
+
+    private val FirSimpleFunction.matchesDataClassSyntheticMemberSignatures: Boolean
+        get() = (this.name == EQUALS_NAME && matchesEqualsSignature) ||
+                (this.name == HASHCODE_NAME && matchesHashCodeSignature) ||
+                (this.name == TOSTRING_NAME && matchesToStringSignature)
```

Comment:
What about `componentX()` and `copy()`?

Reply:
To resolve use sites of those functions, they're added during raw FIR generation, without actual body:
* https://github.com/JetBrains/kotlin/blob/master/compiler/fir/raw-fir/raw-fir.common/src/org/jetbrains/kotlin/fir/builder/BaseFirBuilder.kt#L1078
* https://github.com/JetBrains/kotlin/blob/master/compiler/fir/raw-fir/raw-fir.common/src/org/jetbrains/kotlin/fir/builder/BaseFirBuilder.kt#L1105

Then, their bodies are populated during `fir2ir`. So, from this checker's viewpoint, they should be in the given class's own member declarations already.